### ISSUE TITLE
afl-cmin crashes and pollutes terminal

### ISSUE THREAD

**Issue by fuzzah** [AFLplusplus/AFLplusplus#1428]
afl-cmin fails to minimize huge amount of samples and pollutes terminal with afl-showmap output (raw coverage maps?).

steps to reproduce:
1. fuzz python3 with input scripts (`afl-fuzz ... -- ./python3 -s @@`) for a few days to get bunch of samples.
    in my case i had 8 afl-fuzz instances.
2. put all the samples in a single directory. e.g. `find out -type f -path "*/queue/id*" cp -t samples {} +`
    in my case i had 26136 samples (sizes vary from 1 byte to 30 Kb)
3. run afl-cmin:
    ```
    export AFL_MAP_SIZE=100585 # value reported by calling fuzz target
    afl-cmin -i samples -o samples_cmin -m none -- ./python3 -s @@
    ```

here is what happens in my terminal in a few minutes:
![image](https://user-images.githubusercontent.com/60884276/170672560-8e9a4851-8d50-49ea-898e-b67be006a66b.png)
another example:
![image](https://user-images.githubusercontent.com/60884276/170675396-69c7bb11-504d-48bc-9994-5854380dd417.png)
(note awk error in lower right corner)

exit code of afl-cmin is 2

**Comment by fuzzah** [comment]
i tried piping afl-cmin output to a file.
~90% of the file was the same mess as displayed above.
the last sensible part from afl-cmin is this:
![image](https://user-images.githubusercontent.com/60884276/170685261-b96ab443-8edf-4997-937e-08b9a3829ac0.png)

it says something about failing to compile regexp and something about being out of memory.
the machine has 96 gigs of ram.

**Comment by fuzzah** [comment]
also: afl-cmin logs are extremely uneasy to comprehend because of carriage return symbols.. maybe it should detect that output is being redirected/piped which would mean that inserting \r is not required

**Comment by hexcoder-** [comment]
Thanks for your report!
Could you spare us some days of fuzzing and provide the compressed samples, please?
For comparison and in the mean time you could try the alternative bash script afl-cmin.bash.

**Comment by fuzzah** [comment]
Sure thing @hexcoder- 
[samples_for_issue_1428.tar.gz](https://github.com/AFLplusplus/AFLplusplus/files/8786185/samples_for_issue_1428.tar.gz)

**Comment by fuzzah** [comment]
afl-cmin.bash works here (slowly)

**Comment by hexcoder-** [comment]
Hmm, cannot reproduce. This is how my run went with current dev branch and your samples:

```
heiko@Box:~/AFLplusplus$ afl-cmin -i samples -o samples_cmin -m none -- ./Python-3.10.4/python -s @@
corpus minimization tool for afl++ (awk version)

[*] Testing the target binary...
[!] WARNING: Target reported non-aligned map size of 98692
[+] OK, 9004 tuples recorded.
[*] Obtaining traces for 26136 input files in 'samples'.
    Processing 26136 files (forkserver mode)...
[!] WARNING: Target reported non-aligned map size of 98692
[!] WARNING: Target reported non-aligned map size of 98692
[*] Processing traces for input files in 'samples'.
    Processing file 26136/26136
    Processing tuple 91351/91351 with count 19821...
[+] Found 91351 unique tuples across 26136 files.
[+] Narrowed down to 5365 files, saved in 'samples_cmin'.

```

**Comment by fuzzah** [comment]
@hexcoder- 
i use python:3.9.1-slim-buster docker image, then i install clang, afl++ etc.
tested python version is also 3.9.1, compiled with afl-clang-fast.
afl-cmin versions tested: 3.14c, 4.00c, current dev...

i'll try to create small Dockerfile for reproducing and send it later

**Comment by hexcoder-** [comment]
I compiled the last stable Python version with AFL++ and ran it directly in Ubuntu 18.04...

**Comment by hexcoder-** [comment]
I added redirection detection to the afl-cmin script, but since i am on vacation from tomorrow on, i cannot do more investigation now.

**Comment by hexcoder-** [comment]
Another strange effect I noticed is that results are not reproducible as they should (e.g. number of unique tuples and number of written files).
first run: 
```
...
[+] Found 73348 unique tuples across 26136 files.
[+] Narrowed down to 5684 files, saved in 'samples_cmin'.

```
second run:
```
...
[+] Found 73341 unique tuples across 26136 files.
[+] Narrowed down to 5679 files, saved in 'samples_cmin'.
```

I suspect some problem in afl-showmap...
@vanhauser-thc: could you have a look here please?

**Comment by fuzzah** [comment]
@hexcoder- i guess the reason is non-100% stability of the target.
during my fuzzing the stability was 99%.
upd: those samples were generated with afl-fuzz 3.14c

**Comment by fuzzah** [comment]
i was able to reproduce the issue on **two more** machines using this dockerfile:
[Dockerfile.txt](https://github.com/AFLplusplus/AFLplusplus/files/8789436/Dockerfile.txt)
(need to build and run it without arguments)

**sometimes afl-cmin doesn't crash** and then its output looks like this:
```
[*] Testing the target binary...
[+] OK, 4025 tuples recorded.
[*] Obtaining traces for 26136 input files in 'samples'.
    Processing 26136 files (forkserver mode)...
[*] Processing traces for input files in 'samples'.
    Processing file 26136/26136
    Processing tuple 69980/69980 with count 18829...unknown����8V���8V�{��8V�{��8V�{��8V�{��8V�{��8VX�{��8VS�{��8V�{��8V�{��8V�{��8V▒�{��8V�{��8V▒�{��8V{��8V�{��8V"
[+] Found 69980 unique tuples across 26136 files.
[+] Narrowed down to 6387 files, saved in 'samples_out'.
```

**Comment by fuzzah** [comment]
![image](https://user-images.githubusercontent.com/60884276/170781910-b91f2237-66f1-4906-b1df-c5310b796c76.png)
more pieces of sensible text.......

**Comment by fuzzah** [comment]
i found out that default awk in debian image is actually mawk:
```
# awk -W version
mawk 1.3.3 Nov 1996, Copyright (C) Michael D. Brennan

compiled limits:
max NF             32767
sprintf buffer      2040
```

**switching to gawk seems to mitigate the issue.**

default version in ubuntu:18.04 image is also mawk. @hexcoder- could you please check if you were testing with gawk or mawk?

**Comment by vanhauser-thc** [comment]
@fuzzah can you please recompile the target with current dev state? you are using a target that is compiled with an older state, and we made some breaking changes (and documented that in the changelog). that might already fix your main problem.

the mawk/gawk things needs a look at from our side.

**Comment by fuzzah** [comment]
@vanhauser-thc, to clarify:
originally those samples were generated with afl-fuzz 3.14c and at that time the target was indeed built with afl-cc 3.14c.
in order to test different versions of afl-cmin the target was recompiled with matching versions of afl-cc.

to do so i just changed docker build argument GIT_AFLPP_COMMIT before cloning AFL++ repo (and in turn: before building the target and using afl-cmin). 
upd: talking about this Dockerfile: https://github.com/AFLplusplus/AFLplusplus/issues/1428#issuecomment-1139989213

**Comment by vanhauser-thc** [comment]
the message afl-fuzz/afl-showmap print show that you are using the wrong afl-fuzz/afl-showmap with the target.
```
[!] WARNING: Target reported non-aligned map size of 98692
 ```

**Comment by fuzzah** [comment]
@vanhauser-thc, you're quoting hexcoder's message :) I did not receive such warnings.

**Comment by vanhauser-thc** [comment]
@fuzzah sorry I had no time for this so far and still dont have that :)
to be able to really test this I need to be able to recreate this and I do not know what is needed ...
do you compile+instrument python? if yes, how exactly, with which commands? (wget/git clone, etc. so I can copy + paste)

I downloaded your samples so on that end I am good to go.

thank you!

**Comment by fuzzah** [comment]
@vanhauser-thc please take a look at this gist, it has everything to reproduce the issue:
https://gist.github.com/fuzzah/21a5ede48b9b45884427b60af58e7729
(it's same as in https://github.com/AFLplusplus/AFLplusplus/issues/1428#issuecomment-1139989213)

**Comment by vanhauser-thc** [comment]
ah damn a docker file :)
I am on vacation and I dont have much time here to work on issues plus the Internet is very slow here (like 100kb/s) ... damn this will need to wait until I am back ...

**Comment by vanhauser-thc** [comment]
OK I checked this out after I finally had some time ...
the afl-cmin awk script triggers a bug in mawk where a `print ""` will dump stack or heap memory, and because of control characters printed its all gibberish.
I removed that `print ""` and then for me it works without that weird output being printed now.
the bug seems to come together with too much output it collects for a file descriptor to write into ... because if I remove files from the input directory the issue does not exist either. or when using gawk.
